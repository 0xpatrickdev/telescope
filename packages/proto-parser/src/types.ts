
export interface ProtoDep {
    filename: string;
    package: string;
    imports: string[]
}
export interface ProtoRef {
    absolute: string;
    filename: string;
    proto: ProtoRoot;
}

export interface ProtoRoot {
    package: string;
    imports: string[];
    root: any;
};
