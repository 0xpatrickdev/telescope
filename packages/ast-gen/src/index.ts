export * from './types';
export * from './amino';
export * from './client';
export * from './messages';
export * from './utils';