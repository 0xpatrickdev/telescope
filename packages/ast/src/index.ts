export * from './encoding';
export * from './clients';
export * from './registry';
export * from './services';
export * from './utils';
export * from './bundle';
export * from './types';