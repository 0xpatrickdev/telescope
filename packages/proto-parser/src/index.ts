import { sync as glob } from 'glob';
import { parse, Service } from 'protobufjs';
import { readFileSync } from 'fs';
import { join, resolve as pathResolve } from 'path';
import dotty from 'dotty';

export interface ProtoDep {
    filename: string;
    pkg: string;
    imports: string[]
}

function depResolve(deps: ProtoDep[], protofile: string, resolved: string[], unresolved: string[]) {
    unresolved.push(protofile);

    let dep = deps.find(dep => dep.filename === protofile);
    if (!dep) {
        throw new Error(`Dependency Not Found ${protofile}`)
    }
    let imports = dep.imports ?? [];
    for (let i = 0; i < imports.length; i++) {
        const dep = imports[i];
        if (!resolved.includes(dep)) {
            if (unresolved.includes(dep)) {
                throw new Error(`Circular reference detected ${protofile}, ${dep}`);
            }
            depResolve(deps, dep, resolved, unresolved);
        }
    }
    resolved.push(protofile);
    const index = unresolved.indexOf(protofile);
    unresolved.splice(index);
}

export class ProtoResolver {
    deps: ProtoDep[];
    constructor(deps) {
        this.deps = deps;
    }
    resolve(proto) {
        const resolved = [];
        const unresolved = [];
        depResolve(this.deps, proto, resolved, unresolved);
        return resolved;
    }

    resolveAll() {
        const resolved = [];
        const unresolved = [];


        const allFiles = this.deps.map(({ filename }) => filename)
        const imports = [...new Set(allFiles)];

        const deps: ProtoDep[] = [
            ...this.deps,
            {
                pkg: '_root',
                filename: '_root',
                imports,
            }
        ];
        depResolve(deps, '_root', resolved, unresolved);
        resolved.pop(); // pull _root off
        return resolved;
    }
}

export class ProtoGlobber {
    files: string[];
    protoDir: string;
    deps: ProtoDep[];
    protos: any[];

    constructor(protoDir) {
        this.protoDir = pathResolve(protoDir);
    }

    findProto(filename) {
        return this.getProtos().find(proto => {
            return proto.filename === filename
        });
    }

    getProtoNested(filename) {
        const coinProto = this.findProto(filename);
        const nestedPath = 'root.nested.' + coinProto.proto.package.split('.').join('.nested.') + '.nested';
        return dotty.get(coinProto.proto, nestedPath)
    }

    getProtoNestedObject(filename, objname) {
        const children = this.getProtoNested(filename);
        return children[objname];
    }

    listProtoObjects(filename) {
        const nested = this.getProtoNested(filename);
        return Object.keys(nested);
    }

    listServicesObjects(filename) {
        const nested = this.getProtoNested(filename);
        return Object.keys(nested).map(key => {
            return nested[key]
        }).filter(obj => {
            return obj instanceof Service;
        })
    }

    listServices(filename) {
        const services = this.listServicesObjects(filename);
        return services.map(svc => {
            return {
                name: svc.name,
                methods: Object.keys(svc.methods).map(k => {
                    const method = svc.methods[k];
                    return {
                        name: k,
                        options: method.options,
                        type: method.type,
                        resolved: method.resolved,
                        comment: method.comment,
                        filename: method.filename,
                        requestType: method.requestType,
                        requestStream: method.requestStream,
                        responseType: method.responseType,
                        responseStream: method.responseStream,
                        resolvedRequestType: method.resolvedRequestType,
                        resolvedResponseType: method.resolvedResponseType,
                    }
                })
            }
        })
    }

    getProtoNestedObjectSchema(filename, objname) {
        const obj = this.getProtoNestedObject(filename, objname);
        if (!obj) throw new Error(`cannot find ${objname}`);
        if (!obj.fields) throw new Error(`cannot find ${objname} fields.`);
        const fields = Object.keys(obj.fields).map(key => {
            const field = obj.fields[key];
            return {
                name: field.name,
                type: field.type,
                required: field.required,
                optional: field.optional,
                repeated: field.repeated,
                map: field.map,
                partOf: field.partOf,
                typeDefault: field.typeDefault,
                defaultValue: field.defaultValue,
                long: field.long,
                bytes: field.bytes,
                resolvedType: field.resolvedType,
                extensionField: field.extensionField,
                declaringField: field.declaringField,
                options: field.options
            };
        });
        return {
            name: objname,
            fields
        };
    }

    getProtos() {
        if (this.protos) return this.protos;
        const protoSplat = join(this.protoDir, '/**/*.proto');
        const protoFiles = glob(protoSplat);
        const contents = protoFiles.map(filename => ({
            absolute: filename,
            filename: filename.split(this.protoDir)[1].replace(/^\//, ''),
            content: readFileSync(filename, 'utf-8')
        }))
        const protos = contents.map(({ absolute, filename, content }) => {
            return { proto: parse(content), absolute, filename };
        });
        this.protos = protos;
        return protos;
    }

    getDeps() {
        if (this.deps) return this.deps;
        const deps = this.getProtos().map(el => {
            const {
                filename,
                proto: {
                    package: pkg,
                    imports
                }
            } = el;
            return {
                filename,
                pkg,
                imports
            }
        });
        this.deps = deps;
        return deps;
    }

}
