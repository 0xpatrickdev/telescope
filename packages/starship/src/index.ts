export * from './config';
export * from './utils';
//@ts-ignore
export * from './codegen';
