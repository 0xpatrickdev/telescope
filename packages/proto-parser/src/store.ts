import { sync as glob } from 'glob';
import { parse } from 'protobufjs';
import { readFileSync } from 'fs';
import { join, resolve as pathResolve } from 'path';
import { ProtoDep, ProtoRef } from './types';
import { getNestedProto } from './utils';

export class ProtoStore {
    files: string[];
    protoDir: string;
    deps: ProtoDep[];
    protos: any[];

    constructor(protoDir) {
        this.protoDir = pathResolve(protoDir);
    }

    findProto(filename): ProtoRef {
        return this.getProtos().find(proto => {
            return proto.filename === filename
        });
    }

    findProtoObject(filename, name): any {
        const proto = this.findProto(filename);
        return getNestedProto(proto.proto)[name];
    }

    getProtos(): ProtoRef[] {
        if (this.protos) return this.protos;
        const protoSplat = join(this.protoDir, '/**/*.proto');
        const protoFiles = glob(protoSplat);
        const contents = protoFiles.map(filename => ({
            absolute: filename,
            filename: filename.split(this.protoDir)[1].replace(/^\//, ''),
            content: readFileSync(filename, 'utf-8')
        }))
        const protos = contents.map(({ absolute, filename, content }) => {
            return {
                proto: parse(content, {
                    keepCase: false,
                    alternateCommentMode: true,
                    preferTrailingComment: false
                }), absolute, filename
            };
        });
        this.protos = protos;
        return protos;
    }

    getDeps(): ProtoDep[] {
        if (this.deps) return this.deps;
        const deps = this.getProtos().map(el => {
            const {
                filename,
                proto: {
                    package: pkg,
                    imports
                }
            } = el;
            return {
                filename,
                pkg,
                imports
            }
        });
        this.deps = deps;
        return deps;
    }

}
