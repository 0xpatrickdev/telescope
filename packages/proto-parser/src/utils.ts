import dotty from 'dotty';
import { Service, Type, Enum, Root, Namespace } from 'protobufjs';
import { ProtoStore } from './store';
import { ProtoRef, ProtoRoot } from './types';

export const getNestedProto = (root: ProtoRoot) => {
    const nestedPath = 'root.nested.' + root.package.split('.').join('.nested.') + '.nested';
    return dotty.get(root, nestedPath);
};

export const getServices = (root: ProtoRoot) => {
    const nested = getNestedProto(root);
    return Object.keys(nested).map(key => {
        return nested[key];
    }).filter(obj => {
        return obj instanceof Service;
    }).map(el => {
        return {
            name: el.name,
            ...el.toJSON({ keepComments: true })
        };
    });
};

export const getTypes = (root: ProtoRoot) => {
    const nested = getNestedProto(root);
    return Object.keys(nested).map(key => {
        return nested[key];
    }).filter(obj => {
        return obj instanceof Type;
    }).map(el => {
        return {
            name: el.name,
            ...el.toJSON({ keepComments: true })
        };
    });
};

export const getEnums = (root: ProtoRoot) => {
    const nested = getNestedProto(root);
    return Object.keys(nested).map(key => {
        return nested[key];
    }).filter(obj => {
        return obj instanceof Enum;
    }).map(el => {
        return {
            name: el.name,
            ...el.toJSON({ keepComments: true })
        };
    });
};

export const recursiveLookup = (proto: any, name: string, scope: string[] = [], allowNested = true) => {
    if (!proto) return;

    if (proto.hasOwnProperty(name)) return {
        name,
        scope,
        ...proto[name]
    }

    if (allowNested && proto) {
        const keys = Object.keys(proto);
        for (let k = 0; k < keys.length; k++) {
            const found = recursiveLookup(proto[keys[k]].nested, name, [...scope, keys[k]], allowNested);
            if (found) {
                return {
                    name,
                    ...found
                };
            };
        }
    }
};

/*

  "imports": [
    "google/api/expr/v1alpha1/syntax.proto",
    "google/protobuf/empty.proto",
    "google/protobuf/struct.proto"
  ],

  ...

  "type": "Empty"

  finds Empty inside of the import

*/
export const importLookup = (store: ProtoStore, root: ProtoRoot, name: string) => {
    const objectsFromImports = root.imports.map(imp => {
        const ref = store.findProto(imp);
        return {
            name,
            import: imp,
            importedName: name,
            obj: lookup(store, ref.proto, name, false),
        };
    }).filter(a => !!a.obj);
    if (objectsFromImports.length) return objectsFromImports[0];
};


/*

  "imports": [
    "google/api/expr/v1alpha1/syntax.proto",
    "google/protobuf/empty.proto",
    "google/protobuf/struct.proto"
  ],

  ...

  "type": "google.protobuf.NullValue"

  finds NullValue directly from package name

*/

export const protoImportLookup = (store: ProtoStore, root: ProtoRoot, name: string) => {
    if (name.startsWith('.')) name = name.replace(/^\./, '');
    const nameAsArray = name.split('.');
    const objectName = nameAsArray.pop();
    const packageName = nameAsArray.join('.')
    const objs = store.getProtos()
        // 1 ask all package files of package, e.g "google.protobuf"
        .filter(proto => proto.proto.package === packageName)
        // 2 cross-checking w the imports by filter()
        .filter(proto => root.imports.includes(proto.filename))
        .map((ref: ProtoRef) => {
            return {
                import: ref.filename,
                obj: lookup(store, ref.proto, objectName, false)
            }
        })
        .filter(a => !!a.obj);
    if (objs.length) {
        return {
            import: objs[0].import,
            obj: objs[0].obj,
            importedName: name,
            name: objectName,
            package: packageName
        }
    }
};

export const lookup = (store: ProtoStore, root: ProtoRoot, name: string, allowNested = true) => {
    const nested = getNestedProto(root);
    return recursiveLookup(nested, name, [root.package], allowNested);
};

/*
    nested objects get a slightly different naming convention
    e.g. SignatureDescriptor_Data or SignatureDescriptor_Data_Multi
*/

export const getObjectName = (name: string, scope: string[] = []) => {
    if (!scope.length || scope.length === 1) return name;
    const [_pkg, ...scopes] = scope;
    return [...scopes, name].join('_')
};


export const traverse = (store: ProtoStore, root: ProtoRoot) => {
    const imports = {};

    const obj: ProtoRoot & { parsedImports: Record<string, string[]> } = {
        imports: root.imports,
        package: root.package,
        root: recursiveTraversal(store, root, root.root, imports),
        parsedImports: null
    };
    obj.parsedImports = imports;
    return obj;
};


const NATIVE_TYPES = [
    "double",   // 0
    "float",    // 1
    "int32",    // 2
    "uint32",   // 3
    "sint32",   // 4
    "fixed32",  // 5
    "sfixed32", // 6
    "int64",    // 7
    "uint64",   // 8
    "sint64",   // 9
    "fixed64",  // 10
    "sfixed64", // 11
    "bool",     // 12
    "string",   // 13
    "bytes"     // 14
];


const parseFields = (store: ProtoStore, root: ProtoRoot, obj: any, imports: object) => {
    return Object.keys(obj.fields).reduce((m, key) => {

        const field = obj.fields[key];
        let found: any = null;

        if (NATIVE_TYPES.includes(field.type)) {
            m[key] = {
                parsedType: { name: field.type, type: 'native' },
                isScalar: true,
                typeNum: NATIVE_TYPES.indexOf(field.type),
                ...field.toJSON({ keepComments: true })
            }
            return m;
        }

        found = importLookup(store, root, field.type);
        if (found) {
            imports[found.import] = imports[found.import] || [];
            imports[found.import] = [...new Set([...imports[found.import], found.name])];
            m[key] = {
                parsedType: instanceType(found.obj),
                scopeType: 'import',
                scope: [found.obj.scope],
                ...field.toJSON({ keepComments: true }),
                importedName: found.importedName,
                import: found.import,
            };
            return m;
        }

        found = protoImportLookup(store, root, field.type);
        if (found) {
            imports[found.import] = imports[found.import] || [];
            imports[found.import] = [...new Set([...imports[found.import], found.name])];
            m[key] = {

                parsedType: instanceType(found.obj),
                scopeType: 'protoImport',
                scope: [found.package],
                ...field.toJSON({ keepComments: true }),
                importedName: found.importedName,
                import: found.import,
            };
            return m;
        }

        found = lookup(store, root, field.type);
        if (found) {
            m[key] = {
                scope: found.scope,
                parsedType: instanceType(found),
                ...field.toJSON({ keepComments: true }),
            };
            return m;
        }

        return m;
    }, {});
};

const parseType = (store: ProtoStore, root: ProtoRoot, obj: any, imports: object) => {
    let nested = null;
    if (obj.nested) {
        nested = Object.keys(obj.nested).reduce((m, key) => {
            m[key] = recursiveTraversal(store, root, obj.nested[key], imports);
            return m;
        }, {});
    }
    return {
        type: 'Type',
        name: obj.name,
        options: obj.options,
        fields: parseFields(store, root, obj, imports),
        nested
    }
};

const parseEnum = (store: ProtoStore, root: ProtoRoot, obj: any, imports: object) => {
    return {
        type: 'Enum',
        name: obj.name,
        ...obj.toJSON({ keepComments: true })
    }
};

const parseService = (store: ProtoStore, root: ProtoRoot, obj: any, imports: object) => {
    let nested = null;
    if (obj.nested) {
        nested = Object.keys(obj.nested).reduce((m, key) => {
            m[key] = recursiveTraversal(store, root, obj.nested[key], imports);
            return m;
        }, {});
    }

    return {
        type: 'Service',
        ...obj.toJSON({ keepComments: true }),
        nested
    }
};

export const recursiveTraversal = (store: ProtoStore, root: ProtoRoot, obj: any, imports: object) => {
    if (obj instanceof Type) {
        return parseType(store, root, obj, imports);
    }
    if (obj instanceof Enum) {
        return parseEnum(store, root, obj, imports);
    }
    if (obj instanceof Service) {
        return parseService(store, root, obj, imports);
    }
    if (obj instanceof Root) {
        if (obj.nested) {
            return Object.keys(obj.nested).reduce((m, key) => {
                m.nested[key] = recursiveTraversal(store, root, obj.nested[key], imports);
                return m;
            }, {
                type: 'Root',
                nested: {}
            });
        } else {
            throw new Error('recursiveTraversal() cannot find protobufjs Type')
        }
    }
    if (obj instanceof Namespace) {
        if (obj.nested) {
            return Object.keys(obj.nested).reduce((m, key) => {
                m.nested[key] = recursiveTraversal(store, root, obj.nested[key], imports);
                return m;
            }, {
                type: 'Namespace',
                nested: {}
            });
        } else {
            throw new Error('recursiveTraversal() cannot find protobufjs Type')
        }
    }
    throw new Error('recursiveTraversal() cannot find protobufjs Type')
};

export const instanceType = (obj: any) => {
    if (obj instanceof Type) {
        return {
            name: obj.name,
            type: 'Type'
        };
    }
    if (obj instanceof Enum) {
        return {
            name: obj.name,
            type: 'Enum'
        };
    }
    if (obj instanceof Service) {
        return {
            name: obj.name,
            type: 'Service'
        };
    }
    if (obj instanceof Root) {
        return {
            type: 'Root'
        };
    }
    if (obj instanceof Namespace) {
        return {
            type: 'Namespace'
        };
    }
    // if (obj.name === 'Timestamp') {
    //     return {
    //         name: obj.name,
    //         type: 'google'
    //     };
    // }
    // if (obj.name === 'Duration') {
    //     return {
    //         name: obj.name,
    //         type: 'google'
    //     };
    // }
    if (obj.name.match(/^[a-z]/)) {
        throw new Error('instanceType() cannot find protobufjs Type')
    }
    // duck typing... 
    // TODO why did we loose instance tyeps/names?
    if (obj.fields) {
        return {
            name: obj.name,
            type: 'Type'
        };
    }
    if (obj.values) {
        return {
            name: obj.name,
            type: 'Enum'
        };
    }
    throw new Error('instanceType() cannot find protobufjs Type')
};
