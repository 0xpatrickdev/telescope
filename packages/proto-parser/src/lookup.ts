import { ProtoStore } from './store';
import { ProtoRef, ProtoRoot } from './types';
import { getNestedProto } from './utils';

export const recursiveLookup = (proto: any, name: string, scope: string[] = [], allowNested = true) => {
    if (!proto) return;

    if (proto.hasOwnProperty(name)) return {
        name,
        scope,
        ...proto[name]
    }

    if (allowNested && proto) {
        const keys = Object.keys(proto);
        for (let k = 0; k < keys.length; k++) {
            const found = recursiveLookup(proto[keys[k]].nested, name, [...scope, keys[k]], allowNested);
            if (found) {
                return {
                    name,
                    ...found
                };
            };
        }
    }
};

/*

  "imports": [
    "google/api/expr/v1alpha1/syntax.proto",
    "google/protobuf/empty.proto",
    "google/protobuf/struct.proto"
  ],

  ...

  "type": "Empty"

  finds Empty inside of the import

*/
export const importLookup = (store: ProtoStore, root: ProtoRoot, name: string) => {
    const objectsFromImports = root.imports.map(imp => {
        const ref = store.findProto(imp);
        return {
            name,
            import: imp,
            importedName: name,
            obj: lookup(store, ref.proto, name, false),
        };
    }).filter(a => !!a.obj);
    if (objectsFromImports.length) return objectsFromImports[0];
};


/*

  "imports": [
    "google/api/expr/v1alpha1/syntax.proto",
    "google/protobuf/empty.proto",
    "google/protobuf/struct.proto"
  ],

  ...

  "type": "google.protobuf.NullValue"

  finds NullValue directly from package name

*/

export const protoImportLookup = (store: ProtoStore, root: ProtoRoot, name: string) => {
    if (name.startsWith('.')) name = name.replace(/^\./, '');
    const nameAsArray = name.split('.');
    const objectName = nameAsArray.pop();
    const packageName = nameAsArray.join('.')
    const objs = store.getProtos()
        // 1 ask all package files of package, e.g "google.protobuf"
        .filter(proto => proto.proto.package === packageName)
        // 2 cross-checking w the imports by filter()
        .filter(proto => root.imports.includes(proto.filename))
        .map((ref: ProtoRef) => {
            return {
                import: ref.filename,
                obj: lookup(store, ref.proto, objectName, false)
            }
        })
        .filter(a => !!a.obj);
    if (objs.length) {
        return {
            import: objs[0].import,
            obj: objs[0].obj,
            importedName: name,
            name: objectName,
            package: packageName
        }
    }
};

export const lookup = (store: ProtoStore, root: ProtoRoot, name: string, allowNested = true) => {
    const nested = getNestedProto(root);
    return recursiveLookup(nested, name, [root.package], allowNested);
};
