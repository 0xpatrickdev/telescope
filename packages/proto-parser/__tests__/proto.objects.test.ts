import { ProtoResolver, ProtoGlobber } from '../src/index'
import dotty from 'dotty';
const globber = new ProtoGlobber(__dirname + '/../__fixtures__/chain1');
const resolver = new ProtoResolver(globber.getDeps());
import { Service } from 'protobufjs';

const logstringify = (obj) => console.log(JSON.stringify(obj, null, 2));

it('Coin', () => {
    const objects = globber.listProtoObjects('cosmos/base/v1beta1/coin.proto');
    expect(objects).toEqual([
        "Coin",
        "DecCoin",
        "IntProto",
        "DecProto",
    ]);
});

it('Msg', () => {
    const nested = globber.getProtoNested('osmosis/gamm/v1beta1/tx.proto');
    const objects = globber.listProtoObjects('osmosis/gamm/v1beta1/tx.proto');
    const services = globber.listServicesObjects('osmosis/gamm/v1beta1/tx.proto');
    expect(services.length).toBe(1);
    expect(services[0].name).toEqual('Msg');
});

it('Query', () => {
    const services = globber.listServicesObjects('osmosis/gamm/v1beta1/query.proto');
    // console.log(services);
    expect(services.length).toBe(1);
    expect(services[0].name).toEqual('Query');
});

it('Query listServices', () => {
    const services = globber.listServices('osmosis/gamm/v1beta1/query.proto');
    expect(services).toMatchSnapshot();
});

it('Msg listServices', () => {
    const services = globber.listServices('osmosis/gamm/v1beta1/tx.proto');
    expect(services).toMatchSnapshot();
});
