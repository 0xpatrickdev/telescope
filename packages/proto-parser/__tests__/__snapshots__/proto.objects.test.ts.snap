// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Coin 1`] = `
Array [
  "Coin",
  "DecCoin",
  "IntProto",
  "DecProto",
]
`;

exports[`Msg listServices 1`] = `
Array [
  Object {
    "methods": Array [
      Object {
        "comment": null,
        "filename": null,
        "name": "JoinPool",
        "options": undefined,
        "requestStream": undefined,
        "requestType": "MsgJoinPool",
        "resolved": false,
        "resolvedRequestType": null,
        "resolvedResponseType": null,
        "responseStream": undefined,
        "responseType": "MsgJoinPoolResponse",
        "type": "rpc",
      },
      Object {
        "comment": null,
        "filename": null,
        "name": "ExitPool",
        "options": undefined,
        "requestStream": undefined,
        "requestType": "MsgExitPool",
        "resolved": false,
        "resolvedRequestType": null,
        "resolvedResponseType": null,
        "responseStream": undefined,
        "responseType": "MsgExitPoolResponse",
        "type": "rpc",
      },
      Object {
        "comment": null,
        "filename": null,
        "name": "SwapExactAmountIn",
        "options": undefined,
        "requestStream": undefined,
        "requestType": "MsgSwapExactAmountIn",
        "resolved": false,
        "resolvedRequestType": null,
        "resolvedResponseType": null,
        "responseStream": undefined,
        "responseType": "MsgSwapExactAmountInResponse",
        "type": "rpc",
      },
      Object {
        "comment": null,
        "filename": null,
        "name": "SwapExactAmountOut",
        "options": undefined,
        "requestStream": undefined,
        "requestType": "MsgSwapExactAmountOut",
        "resolved": false,
        "resolvedRequestType": null,
        "resolvedResponseType": null,
        "responseStream": undefined,
        "responseType": "MsgSwapExactAmountOutResponse",
        "type": "rpc",
      },
      Object {
        "comment": null,
        "filename": null,
        "name": "JoinSwapExternAmountIn",
        "options": undefined,
        "requestStream": undefined,
        "requestType": "MsgJoinSwapExternAmountIn",
        "resolved": false,
        "resolvedRequestType": null,
        "resolvedResponseType": null,
        "responseStream": undefined,
        "responseType": "MsgJoinSwapExternAmountInResponse",
        "type": "rpc",
      },
      Object {
        "comment": null,
        "filename": null,
        "name": "JoinSwapShareAmountOut",
        "options": undefined,
        "requestStream": undefined,
        "requestType": "MsgJoinSwapShareAmountOut",
        "resolved": false,
        "resolvedRequestType": null,
        "resolvedResponseType": null,
        "responseStream": undefined,
        "responseType": "MsgJoinSwapShareAmountOutResponse",
        "type": "rpc",
      },
      Object {
        "comment": null,
        "filename": null,
        "name": "ExitSwapExternAmountOut",
        "options": undefined,
        "requestStream": undefined,
        "requestType": "MsgExitSwapExternAmountOut",
        "resolved": false,
        "resolvedRequestType": null,
        "resolvedResponseType": null,
        "responseStream": undefined,
        "responseType": "MsgExitSwapExternAmountOutResponse",
        "type": "rpc",
      },
      Object {
        "comment": null,
        "filename": null,
        "name": "ExitSwapShareAmountIn",
        "options": undefined,
        "requestStream": undefined,
        "requestType": "MsgExitSwapShareAmountIn",
        "resolved": false,
        "resolvedRequestType": null,
        "resolvedResponseType": null,
        "responseStream": undefined,
        "responseType": "MsgExitSwapShareAmountInResponse",
        "type": "rpc",
      },
    ],
    "name": "Msg",
  },
]
`;

exports[`Query listServices 1`] = `
Array [
  Object {
    "methods": Array [
      Object {
        "comment": null,
        "filename": null,
        "name": "Pools",
        "options": Object {
          "(google.api.http).get": "/osmosis/gamm/v1beta1/pools",
        },
        "requestStream": undefined,
        "requestType": "QueryPoolsRequest",
        "resolved": false,
        "resolvedRequestType": null,
        "resolvedResponseType": null,
        "responseStream": undefined,
        "responseType": "QueryPoolsResponse",
        "type": "rpc",
      },
      Object {
        "comment": null,
        "filename": null,
        "name": "NumPools",
        "options": Object {
          "(google.api.http).get": "/osmosis/gamm/v1beta1/num_pools",
        },
        "requestStream": undefined,
        "requestType": "QueryNumPoolsRequest",
        "resolved": false,
        "resolvedRequestType": null,
        "resolvedResponseType": null,
        "responseStream": undefined,
        "responseType": "QueryNumPoolsResponse",
        "type": "rpc",
      },
      Object {
        "comment": null,
        "filename": null,
        "name": "TotalLiquidity",
        "options": Object {
          "(google.api.http).get": "/osmosis/gamm/v1beta1/total_liquidity",
        },
        "requestStream": undefined,
        "requestType": "QueryTotalLiquidityRequest",
        "resolved": false,
        "resolvedRequestType": null,
        "resolvedResponseType": null,
        "responseStream": undefined,
        "responseType": "QueryTotalLiquidityResponse",
        "type": "rpc",
      },
      Object {
        "comment": null,
        "filename": null,
        "name": "Pool",
        "options": Object {
          "(google.api.http).get": "/osmosis/gamm/v1beta1/pools/{poolId}",
        },
        "requestStream": undefined,
        "requestType": "QueryPoolRequest",
        "resolved": false,
        "resolvedRequestType": null,
        "resolvedResponseType": null,
        "responseStream": undefined,
        "responseType": "QueryPoolResponse",
        "type": "rpc",
      },
      Object {
        "comment": null,
        "filename": null,
        "name": "PoolParams",
        "options": Object {
          "(google.api.http).get": "/osmosis/gamm/v1beta1/pools/{poolId}/params",
        },
        "requestStream": undefined,
        "requestType": "QueryPoolParamsRequest",
        "resolved": false,
        "resolvedRequestType": null,
        "resolvedResponseType": null,
        "responseStream": undefined,
        "responseType": "QueryPoolParamsResponse",
        "type": "rpc",
      },
      Object {
        "comment": null,
        "filename": null,
        "name": "TotalPoolLiquidity",
        "options": Object {
          "(google.api.http).get": "/osmosis/gamm/v1beta1/pools/{poolId}/total_pool_liquidity",
        },
        "requestStream": undefined,
        "requestType": "QueryTotalPoolLiquidityRequest",
        "resolved": false,
        "resolvedRequestType": null,
        "resolvedResponseType": null,
        "responseStream": undefined,
        "responseType": "QueryTotalPoolLiquidityResponse",
        "type": "rpc",
      },
      Object {
        "comment": null,
        "filename": null,
        "name": "TotalShares",
        "options": Object {
          "(google.api.http).get": "/osmosis/gamm/v1beta1/pools/{poolId}/total_shares",
        },
        "requestStream": undefined,
        "requestType": "QueryTotalSharesRequest",
        "resolved": false,
        "resolvedRequestType": null,
        "resolvedResponseType": null,
        "responseStream": undefined,
        "responseType": "QueryTotalSharesResponse",
        "type": "rpc",
      },
      Object {
        "comment": null,
        "filename": null,
        "name": "SpotPrice",
        "options": Object {
          "(google.api.http).get": "/osmosis/gamm/v1beta1/pools/{poolId}/prices",
        },
        "requestStream": undefined,
        "requestType": "QuerySpotPriceRequest",
        "resolved": false,
        "resolvedRequestType": null,
        "resolvedResponseType": null,
        "responseStream": undefined,
        "responseType": "QuerySpotPriceResponse",
        "type": "rpc",
      },
      Object {
        "comment": null,
        "filename": null,
        "name": "EstimateSwapExactAmountIn",
        "options": Object {
          "(google.api.http).get": "/osmosis/gamm/v1beta1/{poolId}/estimate/swap_exact_amount_in",
        },
        "requestStream": undefined,
        "requestType": "QuerySwapExactAmountInRequest",
        "resolved": false,
        "resolvedRequestType": null,
        "resolvedResponseType": null,
        "responseStream": undefined,
        "responseType": "QuerySwapExactAmountInResponse",
        "type": "rpc",
      },
      Object {
        "comment": null,
        "filename": null,
        "name": "EstimateSwapExactAmountOut",
        "options": Object {
          "(google.api.http).get": "/osmosis/gamm/v1beta1/{poolId}/estimate/swap_exact_amount_out",
        },
        "requestStream": undefined,
        "requestType": "QuerySwapExactAmountOutRequest",
        "resolved": false,
        "resolvedRequestType": null,
        "resolvedResponseType": null,
        "responseStream": undefined,
        "responseType": "QuerySwapExactAmountOutResponse",
        "type": "rpc",
      },
    ],
    "name": "Query",
  },
]
`;
